### ISSUE TITLE
Race Condition in GoFiber HTML/Template render

### ISSUE THREAD

**Issue by tungdt91** [gofiber/template#116]
Hello Gofiber team,

First of all, thank you all for making a great product.

I have encountered a problem that I believe is very serious with gofiber's html package.

I use the sample of the package at https://github.com/gofiber/template, no problem happens. But when I added the layout path to the first router handle, race conditions occurred.

```
app.Get("/", func(c *fiber.Ctx) error {
		// Render index
		return c.Render("index", fiber.Map{
			"Title": "Hello, World!",
		}, "layouts/main")
})
```

I use the package https://github.com/tsliwowicz/go-wrk to test. The result is as follows:

```
==================
WARNING: DATA RACE
Read at 0x00c0003ae000 by goroutine 16:
  github.com/gofiber/fiber/v2/internal/bytebufferpool.(*ByteBuffer).Write()
      C:/Users/TungDT/go/pkg/mod/github.com/gofiber/fiber/v2@v2.9.0/internal/bytebufferpool/bytebuffer.go:73 +0x45
  text/template.(*state).walk()
      C:/Program Files/Go/src/text/template/exec.go:271 +0x763
  text/template.(*state).walk()
      C:/Program Files/Go/src/text/template/exec.go:264 +0x1ad
  text/template.(*Template).execute()
      C:/Program Files/Go/src/text/template/exec.go:220 +0x2fe
  text/template.(*Template).Execute()
      C:/Program Files/Go/src/text/template/exec.go:203 +0xda
  html/template.(*Template).Execute()
      C:/Program Files/Go/src/html/template/template.go:124 +0x8a
  github.com/gofiber/template/html.(*Engine).Render.func1()
      C:/Users/TungDT/go/pkg/mod/github.com/gofiber/template@v1.6.9/html/html.go:208 +0x70
  runtime.call16()
      C:/Program Files/Go/src/runtime/asm_amd64.s:550 +0x44
  reflect.Value.Call()
      C:/Program Files/Go/src/reflect/value.go:337 +0xe4
  text/template.safeCall()
      C:/Program Files/Go/src/text/template/funcs.go:365 +0xf9
  text/template.(*state).evalCall()
      C:/Program Files/Go/src/text/template/exec.go:725 +0x6ee
  text/template.(*state).evalFunction()
      C:/Program Files/Go/src/text/template/exec.go:580 +0x1de
  text/template.(*state).evalCommand()
      C:/Program Files/Go/src/text/template/exec.go:467 +0x1ac
  text/template.(*state).evalPipeline()
      C:/Program Files/Go/src/text/template/exec.go:436 +0x224
  text/template.(*state).walk()
      C:/Program Files/Go/src/text/template/exec.go:255 +0x571
  text/template.(*state).walk()
      C:/Program Files/Go/src/text/template/exec.go:264 +0x1ad
  text/template.(*Template).execute()
      C:/Program Files/Go/src/text/template/exec.go:220 +0x2fe
  text/template.(*Template).Execute()
      C:/Program Files/Go/src/text/template/exec.go:203 +0xda
  html/template.(*Template).Execute()
      C:/Program Files/Go/src/html/template/template.go:124 +0x8a
  github.com/gofiber/template/html.(*Engine).Render()
      C:/Users/TungDT/go/pkg/mod/github.com/gofiber/template@v1.6.9/html/html.go:211 +0x3a6
  github.com/gofiber/fiber/v2.(*Ctx).Render()
      C:/Users/TungDT/go/pkg/mod/github.com/gofiber/fiber/v2@v2.9.0/ctx.go:875 +0x1cf
  main.main.func1()
      D:/WORK/SelfServe/selfserve.project/test/main.go:24 +0x14e
  github.com/gofiber/fiber/v2.(*App).next()
      C:/Users/TungDT/go/pkg/mod/github.com/gofiber/fiber/v2@v2.9.0/router.go:127 +0x441
  github.com/gofiber/fiber/v2.(*App).handler()
      C:/Users/TungDT/go/pkg/mod/github.com/gofiber/fiber/v2@v2.9.0/router.go:155 +0x1e4
  github.com/gofiber/fiber/v2.(*App).handler-fm()
      C:/Users/TungDT/go/pkg/mod/github.com/gofiber/fiber/v2@v2.9.0/router.go:143 +0x64
  github.com/valyala/fasthttp.(*Server).serveConn()
      C:/Users/TungDT/go/pkg/mod/github.com/valyala/fasthttp@v1.23.0/server.go:2207 +0x1db6
  github.com/valyala/fasthttp.(*Server).serveConn-fm()
      C:/Users/TungDT/go/pkg/mod/github.com/valyala/fasthttp@v1.23.0/server.go:1984 +0x6a
  github.com/valyala/fasthttp.(*workerPool).workerFunc()
      C:/Users/TungDT/go/pkg/mod/github.com/valyala/fasthttp@v1.23.0/workerpool.go:223 +0x10b
  github.com/valyala/fasthttp.(*workerPool).getCh.func1()
      C:/Users/TungDT/go/pkg/mod/github.com/valyala/fasthttp@v1.23.0/workerpool.go:195 +0x49

Previous write at 0x00c0003ae000 by goroutine 17:
  github.com/gofiber/fiber/v2/internal/bytebufferpool.(*ByteBuffer).Write()
      C:/Users/TungDT/go/pkg/mod/github.com/gofiber/fiber/v2@v2.9.0/internal/bytebufferpool/bytebuffer.go:73 +0xc4
  text/template.(*state).walk()
      C:/Program Files/Go/src/text/template/exec.go:271 +0x763
  text/template.(*state).walk()
      C:/Program Files/Go/src/text/template/exec.go:264 +0x1ad
  text/template.(*Template).execute()
      C:/Program Files/Go/src/text/template/exec.go:220 +0x2fe
  text/template.(*Template).Execute()
      C:/Program Files/Go/src/text/template/exec.go:203 +0xda
  html/template.(*Template).Execute()
      C:/Program Files/Go/src/html/template/template.go:124 +0x8a
  github.com/gofiber/template/html.(*Engine).Render()
      C:/Users/TungDT/go/pkg/mod/github.com/gofiber/template@v1.6.9/html/html.go:211 +0x3a6
  github.com/gofiber/fiber/v2.(*Ctx).Render()
      C:/Users/TungDT/go/pkg/mod/github.com/gofiber/fiber/v2@v2.9.0/ctx.go:875 +0x1cf
  main.main.func1()
      D:/WORK/SelfServe/selfserve.project/test/main.go:24 +0x14e
  github.com/gofiber/fiber/v2.(*App).next()
      C:/Users/TungDT/go/pkg/mod/github.com/gofiber/fiber/v2@v2.9.0/router.go:127 +0x441
  github.com/gofiber/fiber/v2.(*App).handler()
      C:/Users/TungDT/go/pkg/mod/github.com/gofiber/fiber/v2@v2.9.0/router.go:155 +0x1e4
  github.com/gofiber/fiber/v2.(*App).handler-fm()
      C:/Users/TungDT/go/pkg/mod/github.com/gofiber/fiber/v2@v2.9.0/router.go:143 +0x64
  github.com/valyala/fasthttp.(*Server).serveConn()
      C:/Users/TungDT/go/pkg/mod/github.com/valyala/fasthttp@v1.23.0/server.go:2207 +0x1db6
  github.com/valyala/fasthttp.(*Server).serveConn-fm()
      C:/Users/TungDT/go/pkg/mod/github.com/valyala/fasthttp@v1.23.0/server.go:1984 +0x6a
  github.com/valyala/fasthttp.(*workerPool).workerFunc()
      C:/Users/TungDT/go/pkg/mod/github.com/valyala/fasthttp@v1.23.0/workerpool.go:223 +0x10b
  github.com/valyala/fasthttp.(*workerPool).getCh.func1()
      C:/Users/TungDT/go/pkg/mod/github.com/valyala/fasthttp@v1.23.0/workerpool.go:195 +0x49

Goroutine 16 (running) created at:
  github.com/valyala/fasthttp.(*workerPool).getCh()
      C:/Users/TungDT/go/pkg/mod/github.com/valyala/fasthttp@v1.23.0/workerpool.go:194 +0x1c4
  github.com/valyala/fasthttp.(*workerPool).Serve()
      C:/Users/TungDT/go/pkg/mod/github.com/valyala/fasthttp@v1.23.0/workerpool.go:147 +0x6a6
  github.com/valyala/fasthttp.(*Server).Serve()
      C:/Users/TungDT/go/pkg/mod/github.com/valyala/fasthttp@v1.23.0/server.go:1737 +0x6b5
  github.com/gofiber/fiber/v2.(*App).Listen()
      C:/Users/TungDT/go/pkg/mod/github.com/gofiber/fiber/v2@v2.9.0/app.go:602 +0x13c
  main.main()
      D:/WORK/SelfServe/selfserve.project/test/main.go:36 +0x211

Goroutine 17 (running) created at:
  github.com/valyala/fasthttp.(*workerPool).getCh()
      C:/Users/TungDT/go/pkg/mod/github.com/valyala/fasthttp@v1.23.0/workerpool.go:194 +0x1c4
  github.com/valyala/fasthttp.(*workerPool).Serve()
      C:/Users/TungDT/go/pkg/mod/github.com/valyala/fasthttp@v1.23.0/workerpool.go:147 +0x6a6
  github.com/valyala/fasthttp.(*Server).Serve()
      C:/Users/TungDT/go/pkg/mod/github.com/valyala/fasthttp@v1.23.0/server.go:1737 +0x6b5
  github.com/gofiber/fiber/v2.(*App).Listen()
      C:/Users/TungDT/go/pkg/mod/github.com/gofiber/fiber/v2@v2.9.0/app.go:602 +0x13c
  main.main()
      D:/WORK/SelfServe/selfserve.project/test/main.go:36 +0x211
```

How do I fix this problem?

Also, is there a way to define the layout path one time only to avoid re-define it again inside Render function?

Thanks!

**Comment by ReneWerner87** [comment]
yeah sure we will look at the report when we find the cause, we will keep you updated

**Comment by tungdt91** [comment]
> yeah sure we will look at the report when we find the cause, we will keep you updated

Thanks for the feedback.
This gives me more confidence when using gofiber.

I wait for your reply.

**Comment by ReneWerner87** [comment]
@tungdt91 have just started with the reproducing of the error

have tested with various tools and can unfortunately find no error
![image](https://user-images.githubusercontent.com/7063188/117997455-406d0780-b343-11eb-82df-cfacd5f5befa.png)
![image](https://user-images.githubusercontent.com/7063188/117997473-4662e880-b343-11eb-83c4-a419c7eb1191.png)
![image](https://user-images.githubusercontent.com/7063188/117997482-4b279c80-b343-11eb-967a-f93ad73264d4.png)


your output looks like it comes from a unittest and not an api test, can you share the code with us, you can also write to me in a direct chat in discord

**Comment by ReneWerner87** [comment]
![image](https://user-images.githubusercontent.com/7063188/117997916-b4a7ab00-b343-11eb-8444-0a6845a59fe4.png)
by the way, i also tested it, as you mentioned in the text above.

**Comment by tungdt91** [comment]
Please run it with command **go run -race main.go** 

Please refer to here:
https://blog.golang.org/race-detector

**Comment by ReneWerner87** [comment]
ok thx
`go run -race main.go`

```go
==================
WARNING: DATA RACE
Read at 0x00c000392018 by goroutine 16:
  github.com/gofiber/fiber/v2/internal/bytebufferpool.(*ByteBuffer).Write()
      /Users/wernerr/projects/testGo/vendor/github.com/gofiber/fiber/v2/internal/bytebufferpool/bytebuffer.go:73 +0x3e
  text/template.(*state).walk()
      /usr/local/Cellar/go/1.16.3/libexec/src/text/template/exec.go:271 +0x747
  text/template.(*state).walk()
      /usr/local/Cellar/go/1.16.3/libexec/src/text/template/exec.go:264 +0x1a4
  text/template.(*state).walkTemplate()
      /usr/local/Cellar/go/1.16.3/libexec/src/text/template/exec.go:418 +0x3cd
  text/template.(*state).walk()
      /usr/local/Cellar/go/1.16.3/libexec/src/text/template/exec.go:269 +0x2e4
  text/template.(*state).walk()
      /usr/local/Cellar/go/1.16.3/libexec/src/text/template/exec.go:264 +0x1a4
  text/template.(*Template).execute()
      /usr/local/Cellar/go/1.16.3/libexec/src/text/template/exec.go:220 +0x2fb
  text/template.(*Template).Execute()
      /usr/local/Cellar/go/1.16.3/libexec/src/text/template/exec.go:203 +0xd2
  html/template.(*Template).Execute()
      /usr/local/Cellar/go/1.16.3/libexec/src/html/template/template.go:124 +0x82
  github.com/gofiber/template/html.(*Engine).Render.func1()
      /Users/wernerr/projects/testGo/vendor/github.com/gofiber/template/html/html.go:208 +0x69
  runtime.call16()
      /usr/local/Cellar/go/1.16.3/libexec/src/runtime/asm_amd64.s:550 +0x3d
  reflect.Value.Call()
      /usr/local/Cellar/go/1.16.3/libexec/src/reflect/value.go:337 +0xd8
  text/template.safeCall()
      /usr/local/Cellar/go/1.16.3/libexec/src/text/template/funcs.go:365 +0xef
  text/template.(*state).evalCall()
      /usr/local/Cellar/go/1.16.3/libexec/src/text/template/exec.go:725 +0x6e4
  text/template.(*state).evalFunction()
      /usr/local/Cellar/go/1.16.3/libexec/src/text/template/exec.go:580 +0x1dc
  text/template.(*state).evalCommand()
      /usr/local/Cellar/go/1.16.3/libexec/src/text/template/exec.go:467 +0x199
  text/template.(*state).evalPipeline()
      /usr/local/Cellar/go/1.16.3/libexec/src/text/template/exec.go:436 +0x207
  text/template.(*state).walk()
      /usr/local/Cellar/go/1.16.3/libexec/src/text/template/exec.go:255 +0x56d
  text/template.(*state).walk()
      /usr/local/Cellar/go/1.16.3/libexec/src/text/template/exec.go:264 +0x1a4
  text/template.(*Template).execute()
      /usr/local/Cellar/go/1.16.3/libexec/src/text/template/exec.go:220 +0x2fb
  text/template.(*Template).Execute()
      /usr/local/Cellar/go/1.16.3/libexec/src/text/template/exec.go:203 +0xd2
  html/template.(*Template).Execute()
      /usr/local/Cellar/go/1.16.3/libexec/src/html/template/template.go:124 +0x82
  github.com/gofiber/template/html.(*Engine).Render()
      /Users/wernerr/projects/testGo/vendor/github.com/gofiber/template/html/html.go:211 +0x3a6
  github.com/gofiber/fiber/v2.(*Ctx).Render()
      /Users/wernerr/projects/testGo/vendor/github.com/gofiber/fiber/v2/ctx.go:875 +0x1cf
  main.main.func2()
      /Users/wernerr/projects/testGo/main.go:29 +0x144
  github.com/gofiber/fiber/v2.(*App).next()
      /Users/wernerr/projects/testGo/vendor/github.com/gofiber/fiber/v2/router.go:127 +0x441
  github.com/gofiber/fiber/v2.(*App).handler()
      /Users/wernerr/projects/testGo/vendor/github.com/gofiber/fiber/v2/router.go:155 +0x1d2
  github.com/gofiber/fiber/v2.(*App).handler-fm()
      /Users/wernerr/projects/testGo/vendor/github.com/gofiber/fiber/v2/router.go:143 +0x54
  github.com/valyala/fasthttp.(*Server).serveConn()
      /Users/wernerr/projects/testGo/vendor/github.com/valyala/fasthttp/server.go:2207 +0x1db6
  github.com/valyala/fasthttp.(*Server).serveConn-fm()
      /Users/wernerr/projects/testGo/vendor/github.com/valyala/fasthttp/server.go:1984 +0x6a
  github.com/valyala/fasthttp.(*workerPool).workerFunc()
      /Users/wernerr/projects/testGo/vendor/github.com/valyala/fasthttp/workerpool.go:223 +0x109
  github.com/valyala/fasthttp.(*workerPool).getCh.func1()
      /Users/wernerr/projects/testGo/vendor/github.com/valyala/fasthttp/workerpool.go:195 +0x44

Previous write at 0x00c000392018 by goroutine 19:
  github.com/gofiber/fiber/v2/internal/bytebufferpool.(*ByteBuffer).Write()
      /Users/wernerr/projects/testGo/vendor/github.com/gofiber/fiber/v2/internal/bytebufferpool/bytebuffer.go:73 +0xc4
  text/template.(*state).walk()
      /usr/local/Cellar/go/1.16.3/libexec/src/text/template/exec.go:271 +0x747
  text/template.(*state).walk()
      /usr/local/Cellar/go/1.16.3/libexec/src/text/template/exec.go:264 +0x1a4
  text/template.(*Template).execute()
      /usr/local/Cellar/go/1.16.3/libexec/src/text/template/exec.go:220 +0x2fb
  text/template.(*Template).Execute()
      /usr/local/Cellar/go/1.16.3/libexec/src/text/template/exec.go:203 +0xd2
  html/template.(*Template).Execute()
      /usr/local/Cellar/go/1.16.3/libexec/src/html/template/template.go:124 +0x82
  github.com/gofiber/template/html.(*Engine).Render()
      /Users/wernerr/projects/testGo/vendor/github.com/gofiber/template/html/html.go:211 +0x3a6
  github.com/gofiber/fiber/v2.(*Ctx).Render()
      /Users/wernerr/projects/testGo/vendor/github.com/gofiber/fiber/v2/ctx.go:875 +0x1cf
  main.main.func2()
      /Users/wernerr/projects/testGo/main.go:29 +0x144
  github.com/gofiber/fiber/v2.(*App).next()
      /Users/wernerr/projects/testGo/vendor/github.com/gofiber/fiber/v2/router.go:127 +0x441
  github.com/gofiber/fiber/v2.(*App).handler()
      /Users/wernerr/projects/testGo/vendor/github.com/gofiber/fiber/v2/router.go:155 +0x1d2
  github.com/gofiber/fiber/v2.(*App).handler-fm()
      /Users/wernerr/projects/testGo/vendor/github.com/gofiber/fiber/v2/router.go:143 +0x54
  github.com/valyala/fasthttp.(*Server).serveConn()
      /Users/wernerr/projects/testGo/vendor/github.com/valyala/fasthttp/server.go:2207 +0x1db6
  github.com/valyala/fasthttp.(*Server).serveConn-fm()
      /Users/wernerr/projects/testGo/vendor/github.com/valyala/fasthttp/server.go:1984 +0x6a
  github.com/valyala/fasthttp.(*workerPool).workerFunc()
      /Users/wernerr/projects/testGo/vendor/github.com/valyala/fasthttp/workerpool.go:223 +0x109
  github.com/valyala/fasthttp.(*workerPool).getCh.func1()
      /Users/wernerr/projects/testGo/vendor/github.com/valyala/fasthttp/workerpool.go:195 +0x44

Goroutine 16 (running) created at:
  github.com/valyala/fasthttp.(*workerPool).getCh()
      /Users/wernerr/projects/testGo/vendor/github.com/valyala/fasthttp/workerpool.go:194 +0x1c4
  github.com/valyala/fasthttp.(*workerPool).Serve()
      /Users/wernerr/projects/testGo/vendor/github.com/valyala/fasthttp/workerpool.go:147 +0x6a6
  github.com/valyala/fasthttp.(*Server).Serve()
      /Users/wernerr/projects/testGo/vendor/github.com/valyala/fasthttp/server.go:1737 +0x6b5
  github.com/gofiber/fiber/v2.(*App).Listen()
      /Users/wernerr/projects/testGo/vendor/github.com/gofiber/fiber/v2/app.go:602 +0x13c
  main.main()
      /Users/wernerr/projects/testGo/main.go:34 +0x211

Goroutine 19 (running) created at:
  github.com/valyala/fasthttp.(*workerPool).getCh()
      /Users/wernerr/projects/testGo/vendor/github.com/valyala/fasthttp/workerpool.go:194 +0x1c4
  github.com/valyala/fasthttp.(*workerPool).Serve()
      /Users/wernerr/projects/testGo/vendor/github.com/valyala/fasthttp/workerpool.go:147 +0x6a6
  github.com/valyala/fasthttp.(*Server).Serve()
      /Users/wernerr/projects/testGo/vendor/github.com/valyala/fasthttp/server.go:1737 +0x6b5
  github.com/gofiber/fiber/v2.(*App).Listen()
      /Users/wernerr/projects/testGo/vendor/github.com/gofiber/fiber/v2/app.go:602 +0x13c
  main.main()
      /Users/wernerr/projects/testGo/main.go:34 +0x211
==================
```

here should be the bug
![image](https://user-images.githubusercontent.com/7063188/118017288-b7f86200-b356-11eb-8e44-9e979543c0b9.png)

but don´t know the solution yet.


@kiyonlin @hi019 any idea?

**Comment by ReneWerner87** [comment]
https://github.com/valyala/bytebufferpool/issues/18

**Comment by tungdt91** [comment]
@ReneWerner87  Sorry for bothering you.
Can we find a solution to this problem?

**Comment by ReneWerner87** [comment]
Sure, if you tell me what exactly the problem is and what the solution to it is, then sure

Currently not sure where the problem is there, the html template engine should be thread safe as well as the byte pool

**Comment by tungdt91** [comment]
> ![image](https://user-images.githubusercontent.com/7063188/117997916-b4a7ab00-b343-11eb-8444-0a6845a59fe4.png)
> by the way, i also tested it, as you mentioned in the text above.

I use it at a basic level, it can be considered that I am using the same as the sample, except that I have many pages and every page I have to use the same layout as your example above.

Unfortunately, I'm also just learning about GO, I'm also trying it out by trial and error but haven't achieved satisfactory results.

I hope the package developers will find the cause for this problem. Meanwhile I will still wait and try to find out the cause.

**Comment by Fenny** [comment]
Thanks for the report @tungdt91, we might need another mutex lock for altering the `lay.Funcs` operation during the execution. @ReneWerner87 and I will start testing tomorrow, will keep you updated 👍

**Comment by tungdt91** [comment]
> Thanks for the report @tungdt91, we might need another mutex lock for altering the `lay.Funcs` operation during the execution. @ReneWerner87 and I will start testing tomorrow, will keep you updated 👍

Great.
I will wait for your update.
Thank you very much.

**Comment by ReneWerner87** [comment]
![image](https://user-images.githubusercontent.com/7063188/120002600-4adf0080-bfd5-11eb-9a6e-509bb5060ae8.png)

this should fix the problem, we just have to check if we don't lose too much performance there

**Comment by tungdt91** [comment]
> ![image](https://user-images.githubusercontent.com/7063188/120002600-4adf0080-bfd5-11eb-9a6e-509bb5060ae8.png)
> 
> this should fix the problem, we just have to check if we don't lose too much performance there

This is great.
Thank you so much for your support.
@ReneWerner87 @Fenny

**Comment by tungdt91** [comment]
By the way, I want to ask more.

Is there a way to configure a default layout without having to pass it multiple times every time the Render function is called?

We can still pass in a layout when we want, but if we don't, the package will receive a default layout when we initialize the View Engine.

**Comment by ReneWerner87** [comment]
fenny found this report in the main repository
https://github.com/golang/go/issues/39807#issuecomment-831609049

last comment is really interesting -> its fixed inside of the package in version 1.17

but i will create a fix for the versions before in a pull request this weekend

**Comment by turgayozgur** [comment]
> ![image](https://user-images.githubusercontent.com/7063188/120002600-4adf0080-bfd5-11eb-9a6e-509bb5060ae8.png)
> 
> this should fix the problem, we just have to check if we don't lose too much performance there

I think that is a bad solution. It blocks all of the requests regardless of the template name. At least, keyed lock or something better should be implemented.

**Comment by ReneWerner87** [comment]
> Is there a way to configure a default layout without having to pass it multiple times every time the Render function is called?

i don't think so, might be worth a pull request, best identical for all enigines

**Comment by ReneWerner87** [comment]
@turgayozgur you can also open a pull request, had read in the article that it can also come with other functions to problems, so I thought mutex lock is the right way, gladly also better 

pull requests are always welcome

**Comment by ReneWerner87** [comment]
for the fix in version 1.17 they also used mutex locks^^
https://go-review.googlesource.com/c/go/+/316669/2/src/text/template/template.go
https://go-review.googlesource.com/c/go/+/316669/2/src/text/template/exec.go


because i think it can also come to problems when users register own functions

**Comment by turgayozgur** [comment]
> for the fix in version 1.17 they also used mutex locks^^
> https://go-review.googlesource.com/c/go/+/316669/2/src/text/template/template.go
> https://go-review.googlesource.com/c/go/+/316669/2/src/text/template/exec.go
> 
> because i think it can also come to problems when users register own functions

Thank you for that detail. RWMutex is also good I think.
